### DISCUSSION TITLE
About CSM Shadow Missing

### DISCUSSION THREAD

**Discussion by JackKaiXing** [google/filament#4814] [Q&A] [answered]
Hello, 

I am working with filament-1.11.0. (same on filament-1.12.11)

When CSM shadow enabled, with camera view changed, part of the shadow from different shadowMap are totally missed.

shadow map for last split: (the far one)
![2021-11-04-4 50 59](https://user-images.githubusercontent.com/21079351/140303081-4579acc8-8ca4-4df7-b6cf-75a417fe3162.png)

shadow map for second split: (the middle one)
![2021-11-04-4 51 19](https://user-images.githubusercontent.com/21079351/140302924-71be4e71-a1e6-4973-a480-79ea7e4ba007.png)

shadow map for first split: (the near one)
![2021-11-04-4 51 33](https://user-images.githubusercontent.com/21079351/140302870-fe1f3233-1e80-4ce5-b99a-0b175cde9cd9.png)

However, on the final rendered image , shadow from the middle and the far are totally missing. Any ideas or which part of code should I explore to find out this possible problem.

![2021-11-04-4 51 52](https://user-images.githubusercontent.com/21079351/140303442-045c2ce7-f1ba-49ab-a704-925d99c23d56.png)

**Comment by JackKaiXing** [comment, marked as answer]
@romainguy 

Besides, for my demo, the stable parameter is default false.

https://github.com/google/filament/blob/a8d3a61c253d677612a42a80fabcddea721e05c3/filament/include/filament/LightManager.h#L270

**Reply by JackKaiXing** [reply]
The below one is expected.
![2021-11-04-4 51 52](https://user-images.githubusercontent.com/21079351/140304897-39be4e51-c288-47d2-8b1a-2cc271efb284.png)

**Reply by JackKaiXing** [reply]
Hello,

I update to filament-1.12.11, this problem still exists. Any ideas?

**Reply by JackKaiXing** [reply]
This problem mostly happens on IOS, sometimes on android.

**Reply by bejado** [reply]
Can you show what the rendered scene looks like? Do you have a sample scene that reproduces the issue? Make sure you're setting the camera's far plane to an appropriate distance.

**Reply by JackKaiXing** [reply]
@bejado  

It could be reproduced using this file. https://github.com/JackKaiXing/glTF/blob/main/testshadow.glb

With the IOS gltf-viewer, just using the above glb and set `ShadowOptions.shadowCascades = 3`, we will get the following results.

`Without CSM:`
![IMG_0488](https://user-images.githubusercontent.com/21079351/140906664-7b0c5fe1-6ce7-4c5f-9d88-514831989b81.PNG)

`With CSM:`
![IMG_0490](https://user-images.githubusercontent.com/21079351/140906819-13ad4420-1825-4cea-94a0-34ca5494ea5a.PNG)
![IMG_0491](https://user-images.githubusercontent.com/21079351/140906829-344b7102-1044-401d-aa94-f97983d2942b.PNG)


```OC
    _indirectLight = IndirectLight::Builder()
                             .reflections(_iblTexture)
                             .irradiance(3, harmonics)
                             .intensity(3000.0f)                                           // here changed
                             .build(*self.modelView.engine);
    self.modelView.scene->setIndirectLight(_indirectLight);

    // Always add a direct light source since it is required for shadowing.
    _sun = EntityManager::get().create();
    LightManager::Builder(LightManager::Type::DIRECTIONAL)
            .color(Color::cct(6500.0f))
            .intensity(10000.0f)
            .direction(math::float3(1.0f, -1.0f, 0.0f))                                          // here changed
            .castShadows(true)
            .build(*self.modelView.engine, _sun);
    self.modelView.scene->addEntity(_sun);
    
    // below added
    auto sun_instance = self.modelView.engine->getLightManager().getInstance(_sun);                    
    auto options = self.modelView.engine->getLightManager().getShadowOptions(sun_instance);   
    options.shadowCascades = 3;                                                                                                                
    self.modelView.engine->getLightManager().setShadowOptions(sun_instance, options);      
```

**Reply by JackKaiXing** [reply]
In the color pass, as the image below, only shadow depth2 has values, while shadow depth1 is empty or dark.

![2021-11-10 10 29 47](https://user-images.githubusercontent.com/21079351/141038706-a94a6217-82ac-4ea2-b7c8-f5ff5d59cbe0.png)
![2021-11-10 10 29 08](https://user-images.githubusercontent.com/21079351/141038721-bb1ea8cc-e5a0-4424-bc29-4e4e92599563.png)

**Reply by JackKaiXing** [reply]
The blue one should be the depth1, i.e. the dark shadow depth map.
![Screen Shot 2021-11-10 at 11 47 56 AM](https://user-images.githubusercontent.com/21079351/141046520-b216265c-91ad-471a-af61-400f6e58f8e2.png)

**Reply by bejado** [reply]
@JackKaiXing I'm unable to reproduce with our gltf-viewer iOS sample set to any number of cascades. Which device are you running on?
![Screen Shot 2021-11-10 at 1 35 24 PM](https://user-images.githubusercontent.com/5298046/141197224-5f886eff-9b2e-4b14-b9d5-85092753d7ce.png)

**Reply by JackKaiXing** [reply]
I tested with two devices.

IPhone X, IOS 15.1
IPhone Xs Max, IOS 14.4.1

Both has that bug.

Mac OS Big Sur 11.5.1
Xcode 13.1

**Reply by JackKaiXing** [reply]
@bejado 

It fails for OpenGL Debug, but works for Metal Debug.

**Reply by bejado** [reply]
Thanks, I do see it now. If this is a bug specific to iOS OpenGL ES, it's possibly a driver issue and unlikely we'll investigate further as we're in the process of deprecating the OpenGL ES backend on iOS. Is there a reason you're unable to use the Metal backend?

**Reply by JackKaiXing** [reply]
@bejado  we could use metal. Thanks.
